### PR TITLE
Ignore more compiler warnings, unify WERROR options

### PR DESCRIPTION
**Pull request by driazati**
This adds some more compiler warnings ignores for everything that happens on a standard CPU build (CUDA builds still have a bunch of warnings so we can't turn on `-Werror` everywhere yet).

Differential Revision: [D28005063](https://our.internmc.facebook.com/intern/diff/28005063/)

### CODE REVIEW

**Comment by ezyang** [comment]
I'm a bit skeptical that this is actually working, given that I don't see any step that adds files to no-werror, or a step that disables the N warnings that are problematic in the codebase. You should at the very least push to a ci-all/ branch to get the full set of configs running. It's also not at all clear to me that you actually succeeded enabling Werror for all CI configs (or was this intentional) and I'm looking for evidence that you actually managed to achieve this.

File: CMakeLists.txt [hunk 1]
```diff
@@ -661,7 +661,11 @@ if(NOT MSVC)
   # Details at http://eigen.tuxfamily.org/bz/show_bug.cgi?id=1459
   string(APPEND CMAKE_CXX_FLAGS " -Wall")
   string(APPEND CMAKE_CXX_FLAGS " -Wextra")
-  string(APPEND CMAKE_CXX_FLAGS " -Werror=return-type")
+  if($ENV{CMAKE_BUILD_WITH_WERROR})
+    string(APPEND CMAKE_CXX_FLAGS " -Werror")
+  else()
+    string(APPEND CMAKE_CXX_FLAGS " -Werror=return-type")
+  endif()
```
Comment (malfet) [inline CMakeLists.txt:668]: Hmm, isn't it already overriding the logic in:
https://github.com/pytorch/pytorch/blob/e8c268746b297efa988e03abc61ff22203bf3980/CMakeLists.txt#L715-L721

Also, it's a bit unconventional to change build settings based on environment variable, adding it as CMAKE option would be a better idea

**Review by malfet — APPROVED** [review]
LGTM, but please address the feature duplication between newly proposed option and one already existing in setup.py and CMakeLists.txt:
https://github.com/pytorch/pytorch/blob/e8c268746b297efa988e03abc61ff22203bf3980/setup.py#L731-L732

**Comment by driazati** [comment]
Moving back to a draft while I address some of @ezyang's concerns and fix these new test failures. #56987 shows that this PR does work on PR builds by removing one of the warnings hidden by this diff, and the tests show an [error](https://app.circleci.com/pipelines/github/pytorch/pytorch/310752/workflows/0b53752d-df70-4e41-8d30-bb7a11f488e9/jobs/12876998) now that the warning has resurfaced.

There are different warnings on binary builds (e.g. [this one](https://app.circleci.com/pipelines/github/pytorch/pytorch/310717/workflows/e982f31a-d069-4997-9fbf-be52e012b514/jobs/12874710)) that make this PR not work as-is, but I think for now this can land with the flag turned off in CI to get the remaining warnings hidden, and in a follow up PR I'll add the big list of files that need to have `-Wno-error` for binary builds + flip the flag.

**Pull request closed without merging**